### PR TITLE
Bug squashing

### PR DESCRIPTION
**Pull request by BenjaminMichaelis**
## Description

Try to fix the bugs in this. Specifically the updating of listings and tests to be correct.

### CODE REVIEW

File: ListingManager/ListingInformation.cs [hunk 1]
```diff
@@ -5,10 +5,13 @@ namespace ListingManager
 {
     public class ListingInformation
     {
+        private string TemporaryExtension = ".tmp";
```
Comment (Keboo) [inline ListingManager/ListingInformation.cs:8]:
```suggestion
        public const string TemporaryExtension = ".tmp";
```

File: ListingManager/ListingInformation.cs [hunk 2]
```diff
@@ -5,10 +5,13 @@ namespace ListingManager
...
         public int ChapterNumber { get; }
         public int ListingNumber { get; }
         public string ListingSuffix { get; }
         public string ListingDescription { get; }
+        public string TemporaryPath { get; }
+        public string Path { get => TemporaryPath.Remove(TemporaryPath.Length - TemporaryExtension.Length, TemporaryExtension.Length); }
```
Comment (Keboo) [inline ListingManager/ListingInformation.cs:14]:
```suggestion
        public string Path => TemporaryPath.Remove(TemporaryPath.Length - TemporaryExtension.Length, TemporaryExtension.Length);
```

File: ListingManager/ListingManager.cs [hunk 3]
```diff
@@ -65,15 +64,48 @@ private static bool TryGetListing(string listingPath, out ListingInformation? li
                     if (result) listingData.Add(data);
                     return result;
                 }).ToList();
+            foreach (string path in allListings)
+            {
+                File.Copy(path, $"{path}.tmp", true);
```
Comment (Keboo) [inline ListingManager/ListingManager.cs:69]:
```suggestion
                File.Copy(path, $"{path}.{ListingInformation.TemporaryExtension}", true);
```

File: ListingManager/ListingManager.cs [hunk 4]
```diff
@@ -65,15 +64,48 @@ private static bool TryGetListing(string listingPath, out ListingInformation? li
...
+                File.Delete(path);
+            }
+            allListings = FileManager.GetAllFilesAtPath($"{pathToChapter}")
+                .OrderBy(x => x)
+                .Where(x => Path.GetExtension(x) == ".tmp").ToList();
```
Comment (Keboo) [inline ListingManager/ListingManager.cs:74]:
```suggestion
                .Where(x => Path.GetExtension(x) == ListingInformation.TemporaryExtension).ToList();
```

File: ListingManager/ListingManager.cs [hunk 5]
```diff
@@ -65,15 +64,48 @@ private static bool TryGetListing(string listingPath, out ListingInformation? li
                     if (result) listingData.Add(data);
                     return result;
                 }).ToList();
+            foreach (string path in allListings)
+            {
+                File.Copy(path, $"{path}.tmp", true);
+                File.Delete(path);
+            }
+            allListings = FileManager.GetAllFilesAtPath($"{pathToChapter}")
```
Comment (Keboo) [inline ListingManager/ListingManager.cs:72]:
```suggestion
            allListings = FileManager.GetAllFilesAtPath(pathToChapter)
```

File: ListingManager/ListingManager.cs [hunk 6]
```diff
@@ -65,15 +64,48 @@ private static bool TryGetListing(string listingPath, out ListingInformation? li
...
+                .OrderBy(x => x)
+                .Where(x =>
+                {
+                    bool result = TryGetListing(x, out var data);
+                    if (result) testListingData.Add(data);
+                    return result;
+                }).ToList();
+            foreach (string path in allTestListings)
+            {
+                File.Copy(path, $"{path}.tmp", true);
```
Comment (Keboo) [inline ListingManager/ListingManager.cs:87]:
```suggestion
                File.Copy(path, $"{path}.{ListingInformation.TemporaryExtension}", true);
```

File: ListingManager/ListingManager.cs [hunk 7]
```diff
@@ -65,15 +64,48 @@ private static bool TryGetListing(string listingPath, out ListingInformation? li
...
+                File.Delete(path);
+            }
+            allTestListings = FileManager.GetAllFilesAtPath($"{pathToChapter}.Tests")
+                .OrderBy(x => x)
+                .Where(x => Path.GetExtension(x) == ".tmp").ToList();
```
Comment (Keboo) [inline ListingManager/ListingManager.cs:92]:
```suggestion
                .Where(x => Path.GetExtension(x) == ListingInformation.TemporaryExtension).ToList();
```

File: ListingManager/ListingManager.cs [hunk 8]
```diff
@@ -65,15 +64,48 @@ private static bool TryGetListing(string listingPath, out ListingInformation? li
...
 
             for (int i = 0, listingNumber = 1; i < allListings.Count; i++, listingNumber++)
             {
+                if (allListings.Count != listingData.Count)
+                {
+                    throw new InvalidOperationException("The number of listing data and allListings doesn't match, possibly .tmp files exist in your directory already.");
```
Comment (Keboo) [inline ListingManager/ListingManager.cs:98]:
```suggestion
                    throw new InvalidOperationException($"The number of listing data and allListings doesn't match, possibly {ListingInformation.TemporaryExtension} files exist in your directory already.");
```

File: ListingManager/ListingManager.cs [hunk 9]
```diff
@@ -140,6 +173,50 @@ private static bool TryGetListing(string listingPath, out ListingInformation? li
...
+                paddedChapterNumber,
+                paddedListingNumber,
+                string.IsNullOrWhiteSpace(listingData) || string.IsNullOrEmpty(listingData) ? "" : $".{listingData}");
+
+            if (verbose)
+            {
+                Console.WriteLine($"Corrective action. {Path.GetFileName(path)} rename to {newFileName}");
+            }
+
+            if (!preview) UpdateNamespaceOfPath(path, newNamespace, newFileName);
```
Comment (Keboo) [inline ListingManager/ListingManager.cs:191]:
```suggestion
            if (!preview)
            {
                UpdateNamespaceOfPath(path, newNamespace, newFileName);
            }
```

File: ListingManager/ListingManager.cs [hunk 10]
```diff
@@ -140,6 +173,50 @@ private static bool TryGetListing(string listingPath, out ListingInformation? li
...
+        /// <param name="curListingData">An instance of ListingInformation for the current listing.</param>
+        private static void UpdateListingNamespace(string path, int chapterNumber, string listingNumber,
+            string listingData, ListingInformation curListingData, bool verbose = false, bool preview = false)
+        {
+            string paddedChapterNumber = chapterNumber.ToString("00");
+
+            string regexSingleDigitListingWithSuffix = @"\d{1}[A-Za-z]";
+            string paddedListingNumber;
+            if (Regex.IsMatch(listingNumber, regexSingleDigitListingWithSuffix))
+            { //allows for keeping the original listing number with a suffix. e.g. "01A"
```
Comment (Keboo) [inline ListingManager/ListingManager.cs:212]: nit:
```suggestion
            {
            //allows for keeping the original listing number with a suffix. e.g. "01A"   
```

File: ListingManager/ListingManager.cs [hunk 11]
```diff
@@ -140,6 +173,50 @@ private static bool TryGetListing(string listingPath, out ListingInformation? li
...
+        /// <param name="listingData">The name of the listing to be included in the namespace/path</param>
+        /// <param name="verbose">When true, enables verbose console output</param>
+        /// <param name="preview">When true, leaves files in place and only print console output</param>
+        /// <param name="curListingData">An instance of ListingInformation for the current listing.</param>
+        private static void UpdateListingNamespace(string path, int chapterNumber, string listingNumber,
+            string listingData, ListingInformation curListingData, bool verbose = false, bool preview = false)
+        {
+            string paddedChapterNumber = chapterNumber.ToString("00");
+
+            string regexSingleDigitListingWithSuffix = @"\d{1}[A-Za-z]";
```
Comment (Keboo) [inline ListingManager/ListingManager.cs:209]:
```suggestion
            const string regexSingleDigitListingWithSuffix = @"\d{1}[A-Za-z]";
```

**Review by Keboo — APPROVED** [review]
Also looks like some NRT warnings that would be worth addressing

**Comment by BenjaminMichaelis** [comment]
Ignoring tests that I broke in the process of fixing the mainly used method, so need to go back later and clean up the tests and other options, but want to get a working version to nuget for mark to use in the meantime.

**Review by Keboo — APPROVED** [review]

**Pull request merged**